controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -1173,6 +1180,64 @@ private long findSegmentSplitsMerges(List<StreamSegmentRecord> referenceSegments
+                    Preconditions.checkArgument(activeTxnRecord.getTxnStatus().equals(TxnStatus.COMMITTING));
+                    if (activeTxnRecord.getCommitOffsets().isEmpty()) {
+                        ActiveTxnRecord updated = new ActiveTxnRecord(activeTxnRecord.getTxCreationTimestamp(),
+                                activeTxnRecord.getLeaseExpiryTime(),
+                                activeTxnRecord.getMaxExecutionExpiryTime(),
+                                TxnStatus.COMMITTING,
+                                activeTxnRecord.getWriterId(),
+                                activeTxnRecord.getCommitTime(),
+                                activeTxnRecord.getCommitOrder(),
+                                ImmutableMap.copyOf(commitOffsets));
+                        return Futures.toVoid(updateActiveTx(epoch, txnId, new VersionedMetadata<>(updated, txnRecord.getVersion())));
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                }), DATA_NOT_FOUND_PREDICATE, null);
+    }
+
+    /**
+     * This method takes the list of transactions in the committing transactions record, and tries to report marks
+     * for writers for these transactions, if they exist. 
+     * WriterId, commit time and commit offsets are recorded in ActiveTxnRecord for each transaction. 
+     * For transactions where above fields are present, a mark is recorded for them. 
+     * This method will ignore any INVALID_TIME or INVALID_POSITION related failures in noting marks for writers.
+     * This is because those will typically arise from idempotent commit case where this and a transaction with higher 
+     * position and time may already have been committed and the overall mark for the writer may already have progressed.
+     * 
+     * @param committingTransactionsRecord Committing transaction record
+     * @return A completableFuture, which when completed will have marks reported for all transactions in the committing 
+     * transaction record for which a writer with time and position information is available. 
+     */
+    CompletableFuture<Void> generateMarksForTransactions(CommittingTransactionsRecord committingTransactionsRecord) {
+        return Futures.allOf(committingTransactionsRecord.getTransactionsToCommit().stream().map(txId -> {
+            int epoch = RecordHelper.getTransactionEpoch(txId);
+            // Ignore data not found exceptions. DataNotFound Exceptions can be thrown because transaction record no longer 
+            // exists and this is an idempotent case. DataNotFound can also be thrown because writer's mark was deleted 
+            // as we attempted to update an existing record. Note: Delete can be triggered by writer explicitly calling
+            // shutdownWriter api. 
+            return Futures.exceptionallyExpecting(getActiveTx(epoch, txId).thenCompose(txnRecord -> {
```

Comment:
I think it would help clarity to have a local variable and do the return Futures.exceptionallyExpecting on the next line.